### PR TITLE
Add support for ITlsConnectionFeature and marshall the APIGW client cert to HttpContext (#786)

### LINKED ISSUE #786: [AspNetCoreServer] When using mutual tls, HttpContext.Request.ConnectionInfo.ClientCertificate is null
A bit of a follow on from #736... while we have the certificate in the `APIGatewayHttpApiV2ProxyRequest` and `APIGatewayProxyRequest` is doesn't flow through to the HttpContext.

## Describe the Feature

Set the HttpContext.Request.ConnectionInfo.ClientCertificate if available on the ProxyRequest.

### Is your Feature Request related to a problem?

Attempting to use sing ClientCertificate when using Mtls with Identity Sever. ( https://identityserver4.readthedocs.io/en/latest/topics/mtls.html )

## Proposed Solution

Add `ItlsConnectionFeature` support to `InvokeFeatures`

### Describe alternatives you've considered

Extract the `ProxyRequest` message from HttpContext.Items and load client cert from there. However this is non-idomatic.

## Additional Context

None

### Environment

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change  

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by damianh**
*Issue #, if available:*

Fixes #786

*Description of changes:*

- Adds `ItlsConnectionFeature` support to `InvokeFeatures`
- Extracts incoming ClientCertificate pem to an `X509Certificate2` and sets the feature property that then flows through to HttpContext
- Works with `APIGatewayHttpApiV2ProxyRequest` and `APIGatewayProxyRequest`
- Added a test and a specific controller to TestWebApp. Works on netcoreapp2.1 and netcoeapp3.1.
- Non-breaking change.

image

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Libraries/src/Amazon.Lambda.AspNetCoreServer/APIGatewayHttpApiV2ProxyFunction.cs [hunk 1]
```diff
@@ -175,6 +176,18 @@ protected override void MarshallRequest(InvokeFeatures features, APIGatewayHttpA
                 // was marshalled into ASP.NET Core request.
                 PostMarshallConnectionFeature(connectionFeatures, apiGatewayRequest, lambdaContext);
             }
+
+            {
+                var tlsConnectionFeature = (ITlsConnectionFeature)features;
+                var clientCertPem = apiGatewayRequest?.RequestContext?.Authentication?.ClientCert?.ClientCertPem;
+                if (clientCertPem != null)
```
Comment (normj) [inline Libraries/src/Amazon.Lambda.AspNetCoreServer/APIGatewayHttpApiV2ProxyFunction.cs:183]: Can you add a verification that the clientCertPem starts with **--------BEGIN CERTIFICATE-----** and **-----END CERTIFICATE-----**? If it doesn't then throw an meaningful error message that the certificate is invalid?

File: Libraries/src/Amazon.Lambda.AspNetCoreServer/APIGatewayProxyFunction.cs [hunk 2]
```diff
@@ -244,6 +245,18 @@ protected override void MarshallRequest(InvokeFeatures features, APIGatewayProxy
...
                 PostMarshallConnectionFeature(connectionFeatures, apiGatewayRequest, lambdaContext);
             }
+
+            {
+                var tlsConnectionFeature = (ITlsConnectionFeature) features;
+                var clientCertPem = apiGatewayRequest?.RequestContext?.Identity?.ClientCert?.ClientCertPem;
+                if (clientCertPem != null)
+                {
+                    // Remove "--------BEGIN CERTIFICATE-----\n" and "-----END CERTIFICATE-----"
+                    clientCertPem = clientCertPem.Substring(28, clientCertPem.Length - 53);
```
Comment (normj) [inline Libraries/src/Amazon.Lambda.AspNetCoreServer/APIGatewayProxyFunction.cs:255]: Same comment as above about verify the cert. Probably put it a utility method that converts incoming cert into the X509Certificate2.

**Review by normj — CHANGES_REQUESTED** [review]
Looks good. Just a minor change request to improve the error handling.

File: Libraries/src/Amazon.Lambda.AspNetCoreServer/APIGatewayHttpApiV2ProxyFunction.cs (hunk 1 above)
Comment (damianh) [reply Libraries/src/Amazon.Lambda.AspNetCoreServer/APIGatewayHttpApiV2ProxyFunction.cs:183]: Added a static method to Internal/Utilities.cs that does this check and the PEM to X509Certificate2 conversion. Updated bit places it's used.

File: Libraries/src/Amazon.Lambda.AspNetCoreServer/APIGatewayProxyFunction.cs (hunk 2 above)
Comment (damianh) [reply Libraries/src/Amazon.Lambda.AspNetCoreServer/APIGatewayProxyFunction.cs]: See comment above https://github.com/aws/aws-lambda-dotnet/pull/787#discussion_r553662430

File: Libraries/src/Amazon.Lambda.AspNetCoreServer/Internal/Utilities.cs [hunk 3]
```diff
@@ -263,5 +264,20 @@ internal static async Task CopyToAsync(Stream source, Stream destination, long?
             // Double-escape any %2F (encoded / characters) so that they survive URL decoding the path.
             .Replace("%2F", "%252F")
             .Replace("%2f", "%252f"));
+
+        internal static X509Certificate2 GetX509Certificate2FromPem(string clientCertPem)
+        {
+            if (!clientCertPem.StartsWith("-----BEGIN CERTIFICATE-----") || clientCertPem.EndsWith("-----END CERTIFICATE-----"))
+            {
+                throw new InvalidOperationException(
```
Comment (damianh) [inline Libraries/src/Amazon.Lambda.AspNetCoreServer/Internal/Utilities.cs:272]: Wasn't sure which Exception to use (please advise...). Didn't add test coverage here, let me know if desired. (TBH, will APIGW ever send an invalid ClientCert PEM?)

**Comment by damianh** [comment]
@normj Changes done as requested. Cheers.

File: Libraries/src/Amazon.Lambda.AspNetCoreServer/Internal/Utilities.cs (hunk 3 above)
Comment (normj) [reply Libraries/src/Amazon.Lambda.AspNetCoreServer/Internal/Utilities.cs:272]: I'm fine with the exception type. Mostly it was the error message I cared about. What I find is people call this library via some unit test with incomplete sample requests. I have learn the hard way that when I give NPE or index out of bound errors for incomplete sample requests it throws people off.

**Review by normj — APPROVED** [review]

**Comment by damianh** [comment]
Thanks for the fast feedback and accepting @normj !

**Comment by normj** [comment]
Thanks again. The PR was just [released today](https://github.com/aws/aws-lambda-dotnet/blob/master/RELEASE.CHANGELOG.md#release-2021-01-11) in version 5.3.0 of Amazon.Lambda.AspNetCoreServer.

**Pull request merged**